plugins/BEdita/Core/src/Model/Table/FoldersTable.php

```diff
@@ -0,0 +1,153 @@
+ *
+ * @since 4.0.0
+ *
+ */
+class FoldersTable extends ObjectsTable
+{
+
+    /**
+     * {@inheritDoc}
+     *
+     * @codeCoverageIgnore
+     */
+    public function initialize(array $config)
+    {
+        parent::initialize($config);
+
+        $this->setEntityClass(Folder::class);
+
+        $this->belongsToMany('Children', [
+            'className' => 'Objects',
+            'through' => 'Trees',
+            'foreignKey' => 'parent_id',
+            'targetForeignKey' => 'object_id',
+            'sort' => [
+                'Trees.tree_left' => 'asc',
+            ],
+        ]);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @codeCoverageIgnore
+     */
+    public function buildRules(RulesChecker $rules)
+    {
+        $rules->add(
+            [$this, 'hasAtMostOneParent'],
+            'hasAtMostOneParent',
+            [
+                'errorField' => 'parents',
+                'message' => __d('bedita', 'Folder can have at most one existing parent.'),
+            ]
+        );
+
+        return $rules;
+    }
+
+    /**
+     * Custom rule for checking that entity has at most one parent.
+     * The check is done on `parents` property
+     *
+     * @param Folder $entity The folder entity to check
+     * @return bool
+     */
+    public function hasAtMostOneParent(Folder $entity)
+    {
+        if (empty($entity->parents)) {
+            return true;
+        }
+
+        $rule = new ValidCount('parents');
+
+        return $rule($entity, ['operator' => '==', 'count' => 1]) && !empty($entity->parent->id);
+    }
+
+    /**
+     * Set `parents` as not dirty to prevent automatic save that could breaks the tree.
+     * The tree is saved later in `afterSave()`
+     *
+     * @param Event $event The event
+     * @param EntityInterface $entity The entity to save
+     * @return void
+     */
+    public function beforeSave(Event $event, EntityInterface $entity)
+    {
+        $entity->dirty('parents', false);
+    }
+
+    /**
+     * Update the tree setting the right parent.
+     *
+     * @param Event $event The event
+     * @param EntityInterface $entity The folder entity persisted
+     * @return void
+     */
+    public function afterSave(Event $event, EntityInterface $entity)
+    {
+        // no update on the tree
+        if (!$entity->isNew() && !$entity->isParentSet()) {
+            return;
+        }
+
+        $trees = TableRegistry::get('Trees');
+
+        if ($entity->isNew()) {
+            $node = $trees->newEntity([
+                'object_id' => $entity->id,
+                'parent_id' => $entity->parent_id,
+            ]);
+            $trees->saveOrFail($node);
+
+            return;
+        }
+
+        $node = $trees->find()
+            ->where(['object_id' => $entity->id])
+            ->firstOrFail();
+
+        // parent unchanged
+        if ($entity->parent_id === $node->parent_id) {
+            return;
+        }
+
+        $node->parent_id = $entity->parent_id;
```

Comment:
Maybe (AKA: maybe not), we could lower number of conditions by doing something nasty like this:

```php
public function afterSave($event, $entity)
{
    $trees = TableRegistry::get('Trees');
    if ($entity->isNew()) {
        // (as above)
    }

    $node = $trees->find()->where([/*...*/])->firstOrFail();
    $node->parent_id = $entity->parent_id; // Same parent_id checks already happens internally.
    if (!$entity->isDirty('parent_node_id')) {
        return; // Tree unchanged.
    }

    $trees->saveOrFail($node);
}
```

This might not be as straightforward as your approach is, but it might save one condition, and perhaps `Folder::isParentSet()` could be removed too.

Reply:
Ok, not too nasty mate, I like it ;)
But...
The first `if` was there to avoid a SELECT on `trees` if no `parent` was set on `Folder` entity. What is better? 
If it's not important I would go with your solution.

Reply:
Uhm… yeah, you're right, I hadn't thought about that. I don't know… we would be saving one `SELECT` only for root folders, which is not a huge optimisation, maybe. Furthermore, if for some reason `$entity->_properties['parent']` has not been loaded, it would fall into that `if` even if it shouldn't.

I 'm not sure which is the better approach…